### PR TITLE
[Autofocus v2] Add Polling To Search Commands for 6.2.0+

### PR DESCRIPTION
**Pull request by DeanArbel**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related issue
fixes https://github.com/demisto/etc/issues/37088
fixes https://github.com/demisto/etc/issues/26464

## Description
1. Added new commands for server versions 6.2.0+: ***autofocus-search-samples-with-results*** and ***autofocus-search-sessions-with-results***.  The new commands allow users to search autofocus with a single command, that doesn't require ***GenericPolling Playbook***. 

2. Fixed an issue where `in progress` searches would return with `complete` result.

3. Changed `alias.hash` queries to `alias.hash_lookup`.

### Addition to playbook
image

## Minimum version of Demisto
- [x] 6.2.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/AutoFocus/ReleaseNotes/1_2_0.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### Palo Alto Networks AutoFocus v2
+- Added new commands for server versions 6.2.0+: ***autofocus-search-samples-with-results*** and ***autofocus-search-sessions-with-results***.
```
Comment (DeanArbel) [inline Packs/AutoFocus/ReleaseNotes/1_2_0.md:4]:
```suggestion
- Added new arguments ***with_results*** to the commands ***autofocus-search-samples*** and ***autofocus-search-sessions***. When ***with_results*** is used, the command will try to return the results.
```

**Comment by DeanArbel** [comment]
@yuvalbenshalom please note this PR is still missing documentation for the new behavior. This will be added once https://github.com/demisto/content-docs/pull/612 is merged.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 2]
```diff
@@ -403,6 +403,26 @@ def ip_to_indicator_type(ip):
             return None
 
 
+class ScheduleMetadata(object):
+    """
```
Comment (anara123) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:407]: I think just to avoid confusion , we delete this class. At the beginning I thought this is the class used by CommandResults.
You use this class only in debug-mode code, for clear coding. I think you can replace it with a function or few lines of code in the debug-mode code.

File: Packs/Base/ReleaseNotes/1_11_0.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
+#### Scripts
+##### CommonServerPython
+Added new class ***Common.PollingConfiguration*** - to be used with ***CommandResults*** to configure the result to poll.
```
Comment (anara123) [inline Packs/Base/ReleaseNotes/1_11_0.md:4]:
```suggestion
Added new class ***Common.ScheduledCommandConfiguration*** - to be used with ***CommandResults*** to configure the result to poll.
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 4]
```diff
@@ -4891,15 +4968,17 @@ class CommandResults:
...
     :return: None
     :rtype: ``None``
     """
 
     def __init__(self, outputs_prefix=None, outputs_key_field=None, outputs=None, indicators=None, readable_output=None,
                  raw_response=None, indicators_timeline=None, indicator=None, ignore_auto_extract=False,
-                 mark_as_note=False, polling_command=None, polling_args=None, polling_timeout=None,
-                 polling_next_run=None, relations=None):
-        # type: (str, object, object, list, str, object, IndicatorsTimeline, Common.Indicator, bool, bool,str, dict, str, str, list) -> None # noqa: E501
+                 mark_as_note=False, relations=None, scheduled_command_config=None):
```
Comment (anara123) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:4980]: prefer short names if it still delivers the message.
`scheduled_command_config` -> `scheduled_command`

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 2 above)
Comment (DeanArbel) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:407]: ok 👍

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 5]
```diff
@@ -1087,6 +1086,51 @@ def test_module():
 
 def search_samples_command():
     args = demisto.args()
+    with_results = args.get('with_results')
```
Comment (anara123) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:1089]: with_results -> polling

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 6]
```diff
@@ -1087,6 +1086,51 @@ def test_module():
...
+        md, info, ctx_prefix, ctx_key = search_samples_helper(args)
+        return_results(CommandResults(outputs=info, readable_output=md, outputs_key_field=ctx_key,
+                                      outputs_prefix=ctx_prefix))
+        return
+    # scheduled
+    Common.ScheduledCommandConfiguration.raise_error_if_not_supported()
+    interval_in_secs = int(args.get('interval_in_seconds', 60))
+    if 'af_cookie' not in args:
+        # start query
+        if 'scope' not in args:
```
Comment (anara123) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:1100]: scope should be required

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 7]
```diff
@@ -1087,6 +1086,51 @@ def test_module():
...
+            raise DemistoException('Please provide a scope for the search.')
+        md, info, ctx_prefix, af_ctx_path = search_samples_helper(args)
```
Comment (anara123) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:1102]: keep search_samples_helper -> search_samples_command
create search_samples_with_polling_command()

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 8]
```diff
@@ -1087,6 +1086,51 @@ def test_module():
...
+        af_cookie = info.get(af_ctx_path)
+        if info.get('Status') != 'complete' and af_cookie:
+            polling_args = {
+                'af_cookie': af_cookie,
+                'interval_in_seconds': interval_in_secs,
+                'with_results': with_results
+            }
+            polling_config = Common.ScheduledCommandConfiguration(command='autofocus-search-samples',
```
Comment (anara123) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:1110]: Common.ScheduledCommandConfiguration -> ScheduledCommand

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 9]
```diff
@@ -1137,17 +1218,15 @@ def search_sessions_command():
...
-    })
+    return md, info, 'AutoFocus.SessionsSearch', 'AFCookie'
 
 
 def samples_search_results_command():
     args = demisto.args()
+    samples_search_results_helper(args)
+
+
+def samples_search_results_helper(args):
```
Comment (anara123) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:1229]: revert this change. keep the name `def samples_search_results_command(args):`

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 10]
```diff
@@ -1104,17 +1148,54 @@ def search_samples_command():
...
-    })
+    return md, info, 'AutoFocus.SamplesSearch', 'AFCookie'
 
 
 def search_sessions_command():
     args = demisto.args()
+    with_results = args.get('with_results')
+    if with_results != 'true':
+        md, info, ctx_prefix, ctx_key = search_sessions_helper(args)
+        return_results(CommandResults(outputs=info, readable_output=md, outputs_key_field=ctx_key,
```
Comment (anara123) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:1159]: call `return_results` in the main.
in the commands return CommandResults.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 11]
```diff
@@ -4745,6 +4769,64 @@ def to_context(self):
...
+    :type args: ``Optional[Dict[str, Any]]``
+    :param args: Arguments to use when executing the command.
+
+    :type timeout_in_seconds: ``Optional[int]``
+    :param timeout_in_seconds: Number of seconds until the polling sequence will timeout.
+
+    :return: None
+    :rtype: ``None``
+    """
+    VERSION_MISMATCH_ERROR = 'This command is not supported for your server version. Please update your server ' \
```
Comment (yuvalbenshalom) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:4792]:
```suggestion
    VERSION_MISMATCH_ERROR = 'This command is not supported by this XSOAR server version. Please update your server ' \
```

**Review by yuvalbenshalom — COMMENTED** [review]
very nice.
shouldn't this include also a readme update?

**Comment by DeanArbel** [comment]
> very nice.
> shouldn't this include also a readme update?

it should. will be taken care of.

**Review by yuvalbenshalom — APPROVED** [review]

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 12]
```diff
@@ -5414,10 +5499,8 @@ def __init__(self, outputs_prefix=None, outputs_key_field=None, outputs=None, in
         self.indicators_timeline = indicators_timeline
         self.ignore_auto_extract = ignore_auto_extract
         self.mark_as_note = mark_as_note
-        self.polling_command = polling_command
-        self.polling_args = polling_args
-        self.polling_timeout = polling_timeout
-        self.polling_next_run = polling_next_run
+        self.scheduled_command = scheduled_command
+        self.disable_readable_output = disable_readable_output
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:5503]: Actually, in these cases, it does return an output, but since it won't contain a true update, I can also just remove the output and deal with it as you suggested 👍

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 13]
```diff
@@ -259,6 +259,54 @@
...
+        }
+        schedule_config = ScheduledCommand(command=cmd,
+                                           next_run_in_seconds=interval_in_secs,
+                                           args=polling_args, timeout_in_seconds=600)
+        if command_results is None:
+            command_results = CommandResults(scheduled_command=schedule_config, disable_readable_output=True)
+        elif isinstance(command_results, list):
+            command_results[0].scheduled_command = schedule_config
+            for result in command_results:
+                result.disable_readable_output = True
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:303]: For the tags, but I looked at the code again and it can be handled without a list, I'll remove this handling.

File: Packs/AutoFocus/TestPlaybooks/playbook-AutoFocus_V2_test.yml [hunk 14]
```diff
@@ -2202,19 +2173,567 @@ tasks:
...
+    isoversize: false
+    isautoswitchedtoquietmode: false
+  "54":
+    id: "54"
+    taskid: cc33affb-d0a1-4351-88f5-7a169f234a48
+    type: condition
+    task:
+      id: cc33affb-d0a1-4351-88f5-7a169f234a48
+      version: -1
+      name: Is Demisto Version 6.2.0+
```
Comment (yuvalbenshalom) [inline Packs/AutoFocus/TestPlaybooks/playbook-AutoFocus_V2_test.yml:2216]: nice

**Review by yuvalbenshalom — COMMENTED** [review]
good work.
